### ISSUE TITLE
Calling Role from another Role should be more intuitive

### ISSUE THREAD

**Issue by rcschrg** [OFFIS-DAI/mango#52]
In GitLab by @maurerle on Apr 19, 2023, 17:10

Currently, there is no easy way to "notify" a Role of the same Agent in Mango.
This can only be done when messaging the Agent itself.

Other options would include having a Future in the data context, which is not that clean either.
It might be useful to create an example of this interaction between Roles of the same Agent

**Comment by rcschrg** [comment]
@maurerle @towolff @jsagerOffis 
I would appreciate your opinions on this. I started to implement an event system to emit events from role to notify other roles. I think there are some open questions:
1. sync or async (currently sync)?
2. with or without direct answers (e.g. event return values)
3. notify indirect or direct (should it be possible emitting events to specific roles, role types)

The current draft look like this
```python
class Event ...

class RoleHandlingEvents(Role):
    def __init__(self):
        self.source = None
        self.event = None

    def handle_event(self, event, source):
        self.event = event
        self.source = source

    def setup(self) -> None:
        self.context.subscribe_event(self, Event, self.handle_event)

class OtherRole(Role):
    def setup(self) -> None:
        self.context.emit_event(Event("Here we are!")) # event_source is per default None
```

Thanks for your feedback!

**Comment by towolff** [comment]
Nice! 

Regarding your questions:
1. I don´t see a reason for async here
2. At least the way I would initially use events, I wouldn't need any direct responses at first, as I would only use the event to trigger an activity in another role. If the result of this activity is to come back to the event-triggering role, I would probably do this by I.) triggering a new event in the recipient role or II.) sending a message to the triggering role. Whereby II mixes two approaches (sending messages and triggering events). So: what would be the best approach for "inter-role activity triggering"? -> Only via Events? 
3. I think I didn't really understand this point.

**Comment by rcschrg** [comment]
Regarding 1. I feel there could be a reason to handle an event async, but I also don't see a reason to send it async.
Regarding 2. In my opinion there shouldn't be a reason for an immediate answer, if you need to return a result it could happen using models or the data container, or even a new event.
Clarifying 3. Should it be possible to emit an event to a specific role? (e.g. emit_event(MyTargetRole, Event()).

**Comment by maurerle** [comment]
> I feel there could be a reason to handle an event async, but I also don't see a reason to send it async.
True, but I think that sync should be enough until we encounter this edge case

>  what would be the best approach for "inter-role activity triggering"? -> Only via Events?
Yes I think if you want to communicate inside a role, having it only through events would be best

> Should it be possible to emit an event to a specific role? (e.g. emit_event(MyTargetRole, Event()).
I think that it is flexible enough to check the EventType as in your PR

**Comment by rcschrg** [comment]
was fixed with #65
